## Initial Data Diagnosis

In [16]:
import numpy as np
import pandas as pd
SEED = 39
np.random.seed(SEED)


df = pd.read_csv('creditcard.csv')

In [17]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [18]:
df.describe()

Time            V1            V2            V3            V4  \
count  284807.000000  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean    94813.859575  1.175161e-15  3.384974e-16 -1.379537e-15  2.094852e-15   
std     47488.145955  1.958696e+00  1.651309e+00  1.516255e+00  1.415869e+00   
min         0.000000 -5.640751e+01 -7.271573e+01 -4.832559e+01 -5.683171e+00   
25%     54201.500000 -9.203734e-01 -5.985499e-01 -8.903648e-01 -8.486401e-01   
50%     84692.000000  1.810880e-02  6.548556e-02  1.798463e-01 -1.984653e-02   
75%    139320.500000  1.315642e+00  8.037239e-01  1.027196e+00  7.433413e-01   
max    172792.000000  2.454930e+00  2.205773e+01  9.382558e+00  1.687534e+01   

                 V5            V6            V7            V8            V9  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   1.021879e-15  1.494498e-15 -5.620335e-16  1.149614e-16 -2.414189e-15   
std    1.380247e+00  1.332271e+00  1.237094e+00  1.194353e+00  1.098632e+00   
min   -1.137433e+02 -2.616051e+01 -4.355724e+01 -7.321672e+01 -1.343407e+01   
25%   -6.915971e-01 -7.682956e-01 -5.540759e-01 -2.086297e-01 -6.430976e-01   
50%   -5.433583e-02 -2.741871e-01  4.010308e-02  2.235804e-02 -5.142873e-02   
75%    6.119264e-01  3.985649e-01  5.704361e-01  3.273459e-01  5.971390e-01   
max    3.480167e+01  7.330163e+01  1.205895e+02  2.000721e+01  1.559499e+01   

       ...           V21           V22           V23           V24  \
count  ...  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05   
mean   ...  1.628620e-16 -3.576577e-16  2.618565e-16  4.473914e-15   
std    ...  7.345240e-01  7.257016e-01  6.244603e-01  6.056471e-01   
min    ... -3.483038e+01 -1.093314e+01 -4.480774e+01 -2.836627e+00   
25%    ... -2.283949e-01 -5.423504e-01 -1.618463e-01 -3.545861e-01   
50%    ... -2.945017e-02  6.781943e-03 -1.119293e-02  4.097606e-02   
75%    ...  1.863772e-01  5.285536e-01  1.476421e-01  4.395266e-01   
max    ...  2.720284e+01  1.050309e+01  2.252841e+01  4.584549e+00   

                V25           V26           V27           V28         Amount  \
count  2.848070e+05  2.848070e+05  2.848070e+05  2.848070e+05  284807.000000   
mean   5.109395e-16  1.686100e-15 -3.661401e-16 -1.227452e-16      88.349619   
std    5.212781e-01  4.822270e-01  4.036325e-01  3.300833e-01     250.120109   
min   -1.029540e+01 -2.604551e+00 -2.256568e+01 -1.543008e+01       0.000000   
25%   -3.171451e-01 -3.269839e-01 -7.083953e-02 -5.295979e-02       5.600000   
50%    1.659350e-02 -5.213911e-02  1.342146e-03  1.124383e-02      22.000000   
75%    3.507156e-01  2.409522e-01  9.104512e-02  7.827995e-02      77.165000   
max    7.519589e+00  3.517346e+00  3.161220e+01  3.384781e+01   25691.160000   

               Class  
count  284807.000000  
mean        0.001727  
std         0.041527  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         0.000000  
max         1.000000  

[8 rows x 31 columns]

- Dataset consists of 31 quantititave variables, with the target being "Class". Notably, variables V1, V2, ..., V28 are principle components obtained with PCA already while feature "Amount" was kept for interpretation, which might introduce redundant scaling bias into the static model. Standardization for that particular feature might be needed to mitigate this.  

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

- Data is efficiently stored in float64 and int64 values, taking up 67.4 MB

In [20]:
df.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [21]:
print(df.Class.value_counts(normalize=True) * 100)
print(df.duplicated().sum())

Class
0    99.827251
1     0.172749
Name: proportion, dtype: float64
1081


- As seen from this proportion, the positive events are extremely dominated by the negative events. To address this, methods like class weights and SMOTE might be implemented.
- No null values but some duplications.

## Data Cleaning

### Duplication

- As observed in the Data Diagnosis phase, there are some duplicates in this dataset of credit card transaction. This is common as credit card holders can sometimes swipe twice or encounter errors with the credit card processing.

In [22]:
df.drop_duplicates(inplace=True)
df.drop(columns=["Time"],inplace=True)
df.shape

(283726, 30)

- Drop identical duplicates for reduced dataset bias and column "Time" to focus on static modeling. 

## Data Preprocessing

### Data Splitting, Stratefied Cross Validation

- When developing a ML model, dedicating partitions of the dataset exclusively for training and testing provides an opportunity to observe how well the model would perform with unseen real data, if deployed in real world.
- Within the training processes, we further implement cross validation to ensure our trained model is not overfitting or underfitting to our limited training data. Stratefied Cross Validation is utilized to preserve class ratio given the imbalanced dataset.
- For scoring metrics, we refer to threshold-free metrics like ROC AUC and PR AUC for testing the hypotheses in the Data Preprocessing Phase as the final threshold hasn't been decided yet. Had we chose metrics from the confusion matrix of the default threshold 0.5, we can be completely biased into designing a pipeline that optimizes for the threshold 0.5 only, not all of the possible values.
    - As mentioned, due to the imbalance nature of this problem, a high ROC AUC score can be misleading as a low FPR can still indicate massive false positives due to the dominant negative class. This leads us to the Precision-Recall curve, which put a constraint context on minimizing type 1 and type 2 errors with respect to the true positives, so that the false positives doesn't blow out of proportion.
    - Ideally, we want to maximize PR AUC score in this problem so that during the model selection phase, we can fix a constraint for recall and maximize precision.

In [23]:
X = df.drop(columns='Class')
y = df['Class']

from sklearn.model_selection import train_test_split, StratifiedKFold
from sklearn.metrics import make_scorer, average_precision_score

# Splitting the training/testing datasets
X_train, X_test, y_train, y_test = train_test_split(X, 
                                                    y, 
                                                    test_size=0.2, 
                                                    stratify=y, 
                                                    random_state=SEED)
# Define Stratified Cross Validation
cv = StratifiedKFold(n_splits=5, 
                     shuffle=True, 
                     random_state=SEED)

# Define scoring metrics
scoring_metrics = {
    'pr_auc': make_scorer(average_precision_score),
}

# Boilerplate code for displaying metrics
def print_results(cv_results):
    print("Out-of-fold predictions:")
    print(f"Avg PR AUC: {np.mean(cv_results['test_pr_auc']):.4f} (+/- {np.std(cv_results['test_pr_auc']):.4f})")

    print("\nIn-fold predictions:")
    print(f"Avg PR AUC: {np.mean(cv_results['train_pr_auc']):.4f} (+/- {np.std(cv_results['train_pr_auc']):.4f})")


### Scaling - Standardization

TODO:
- Hypothesis: Other scaling methods might outperform standardization for this model.

- As mentioned above, standardization for feature "Amount" helps reduce model's bias and optimize the model fitting process.

In [24]:
from sklearn.pipeline import Pipeline
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import cross_validate

ct = ColumnTransformer(
    transformers=[
        ('scale_amount', StandardScaler(), ['Amount']) # 
    ],
    remainder='passthrough'  # keep all other columns untouched
)


# Define base model pipeline with StandardScaler
pipe = Pipeline([
    ('scaler', ct), # apply custom scaler
    ('model', LogisticRegression(max_iter=1000)) # bypass warnings for default configure
])

# Obtain cross validation results
cv_results = cross_validate(pipe, 
                            X_train, 
                            y_train, 
                            cv=cv, 
                            scoring=scoring_metrics,
                            return_train_score=True)

# Print the results
print_results(cv_results)

Out-of-fold predictions:
Avg PR AUC: 0.5340 (+/- 0.0334)

In-fold predictions:
Avg PR AUC: 0.5462 (+/- 0.0221)


### Outliers

TODO:
- Hypothesis: Removing outliers in the majority class only might improve the PR curve.

- No outliers detection or removal was done to preserve the scarce imbalanced data in this dataset. Any datapoint, may it be positive or negative, is crucial to the model performance.

### Class Imbalance

In [25]:
pipe_class_weight = Pipeline([
    ('scaler', ct),
    ('model', LogisticRegression(max_iter=1000, 
                                 class_weight='balanced'))
])

cv_results = cross_validate(pipe_class_weight, 
                            X_train, 
                            y_train, 
                            cv=cv, 
                            scoring=scoring_metrics,
                            return_train_score=True)

print_results(cv_results)

Out-of-fold predictions:
Avg PR AUC: 0.0524 (+/- 0.0027)

In-fold predictions:
Avg PR AUC: 0.0543 (+/- 0.0037)


In [30]:
from imblearn.pipeline import Pipeline as imbPipeline 
from imblearn.over_sampling import SMOTE

pipe_smote = imbPipeline([
    ('scaler', ct),
    ('sampler', SMOTE(random_state=SEED)), 
    ('model', LogisticRegression(max_iter=1000))
])

cv_results = cross_validate(pipe_smote, 
                            X_train, 
                            y_train, 
                            cv=cv, 
                            scoring=scoring_metrics,
                            return_train_score=True)

print_results(cv_results)

Out-of-fold predictions:
Avg PR AUC: 0.0496 (+/- 0.0019)

In-fold predictions:
Avg PR AUC: 0.0516 (+/- 0.0035)


- As seen from these results, implementing imbalance methods, like Class-Weighting and SMOTE, aren't effective for maximizing PR AUC for this Fraud Detection System. The main reason being that these methods improve the ROC AUC but also simultaneously flag a lot of nearby positive outliers to be positive as well, hence the abundance of false positives.

### Polynomial and Interaction Features

TODO:
- Hypothesis: To maximize model performance while maintaining the algorithm's computing feasibility, the optimal maximum degree of polynomial is 2.
    - Comment: 2 is enough to ensure model complexity is causing overfitting, with 86% average ROC AUC score for the training folds.

In [29]:
from sklearn.preprocessing import PolynomialFeatures

pipe_poly = Pipeline([
    ('scaler', ct),
    ('poly', PolynomialFeatures(degree=2, 
                                include_bias=False)),
    ('model', LogisticRegression(max_iter=1000))
])

cv_results = cross_validate(pipe_poly, 
                            X_train, 
                            y_train, 
                            cv=cv, 
                            scoring=scoring_metrics,
                            return_train_score=True)

print_results(cv_results)

Out-of-fold predictions:
Avg PR AUC: 0.6325 (+/- 0.0560)

In-fold predictions:
Avg PR AUC: 0.8583 (+/- 0.0072)


- As expected, adding polynomial features increases the model complexity but it's also prone to overfitting. The next steps in the model selection phase implement regularization and hyperparameter tuning to address this.

## Hyperparameter Tuning

### Regularization

TODO:
- Hypothesis: Is grid search the most optimal algorithm?

In [33]:
def print_grid_search_results(gs_results):
    best_index = gs_results.best_index_
    cv_results = gs_results.cv_results_
    print("Out-of-fold predictions (for best parameter):")
    for scorer in gs_results.scorer_:
        print(f"Avg {scorer}: \t{cv_results[f'mean_test_{scorer}'][best_index]:.4f} (+/- {cv_results[f'std_test_{scorer}'][best_index]:.4f})")

    print("\nIn-fold predictions (for best parameter):")
    for scorer in gs_results.scorer_:
        print(f"Avg {scorer}: \t{cv_results[f'mean_train_{scorer}'][best_index]:.4f} (+/- {cv_results[f'std_train_{scorer}'][best_index]:.4f})")

In [42]:
from sklearn.model_selection import GridSearchCV

pipe_poly = Pipeline([
    ('scaler', ct),
    ('poly', PolynomialFeatures(degree=2, 
                                include_bias=False)),
    ('model', LogisticRegression(max_iter=1000,
                                 penalty='l2'))
])

param_grid = {
    'model__C': np.logspace(-3, 2, 6)
}

grid_search = GridSearchCV(
    estimator=pipe_poly,
    param_grid=param_grid,
    scoring=scoring_metrics,
    refit='pr_auc',
    cv=cv,
    return_train_score=True,
    verbose=1,
    n_jobs=10 # Depends on your machine RAM, adjust this number accordingly
)

grid_search.fit(X_train, y_train)

print(f"Best C parameter: {grid_search.best_params_}")
print(f"Best PR AUC score: {grid_search.best_score_:.4f}\n")

print_grid_search_results(grid_search)

Fitting 5 folds for each of 6 candidates, totalling 30 fits
Best C parameter: {'model__C': np.float64(0.01)}
Best PR AUC score: 0.6450
Out-of-fold predictions (for best parameter):
Avg pr_auc: 	0.6450 (+/- 0.0405)

In-fold predictions (for best parameter):
Avg pr_auc: 	0.7790 (+/- 0.0065)


- It seems that for Logistic Regression, this value of PR AUC of 64.5% is approaching the upper limit of the model capabilities. We can finalize the model by finding the best threshold next.

### Threshold Tuning

In [58]:
from sklearn.metrics import precision_recall_curve
from sklearn.model_selection import cross_val_predict

BETA = 3  # Adjust this constant to emphasize recall better

best_c = LogisticRegression(max_iter=1000,
                            penalty='l2',
                            C=0.01)

y_cv_scores = cross_val_predict(best_c, 
                                X_train, 
                                y_train, 
                                cv=cv,
                                method="predict_proba")[:,1]

precision, recall, thresholds = precision_recall_curve(y_train, y_cv_scores)
p = precision[:-1]
r = recall[:-1]

fbeta_scores = (1 + BETA**2) * (p * r) / ((BETA**2 * p) + r)
fbeta_scores = np.nan_to_num(fbeta_scores, nan=0.0)

best_idx = np.argmax(fbeta_scores)
best_threshold = thresholds[best_idx]
best_fbeta = fbeta_scores[best_idx]
best_precision = p[best_idx]
best_recall = r[best_idx]

print(f"Optimal Threshold (from CV): {best_threshold:.4f}")
print(f"Max F{BETA}-Score (on CV):    {best_fbeta:.4f}")
print(f"Precision at Threshold (on CV): {best_precision:.4f}")
print(f"Recall at Threshold (on CV):  {best_recall:.4f}")

Optimal Threshold (from CV): 0.0176
Max F3-Score (on CV):    0.8248
Precision at Threshold (on CV): 0.6414
Recall at Threshold (on CV):  0.8519


## Model Evaluation